### PR TITLE
Fix search relevance wrong sort order

### PR DESCRIPTION
**Pull request by Kershoc**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       |  develop
| Description?  | BOOM-3915 Relevance search wrong sort order
| Type?         | bug fix 
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-3915
| How to test?  | Search for "printed dress" in the demo shop.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8399)

### CODE REVIEW

File: controllers/front/listing/SearchController.php [hunk 1]
```diff
@@ -66,7 +66,7 @@ protected function getProductSearchQuery()
     {
         $query = new ProductSearchQuery();
         $query
-            ->setSortOrder(new SortOrder('product', Tools::getProductsOrder('by'), Tools::getProductsOrder('way')))
+            ->setSortOrder(new SortOrder('product', Tools::getProductsOrder('by'), 'desc'))
```
Comment (Quetzacoalt91) [inline controllers/front/listing/SearchController.php]: It seems you are removing the sorting feature from the search by forcing a value. Are you sure about this change?

Comment (Kershoc) [reply controllers/front/listing/SearchController.php]: Yes you are correct.  That file was not suppose to make it into the commit.  This is what I get for rushing.  I'll amend this commit to remove this change.  My apologies.

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by sbordun** [comment]
HI,

I tested the change and it is working, but you forgot the Ajax search from the header tool bar.
It shows results immediately but in the ASC relevance order instead of DESC.

Could I suggest additional change that I tested and it is working fine on my site?

controllers/front/listing/SearchController

     protected function getProductSearchQuery()
     {
.....................

REMOVE THIS LINE
          ->setSortOrder(new SortOrder('product', Tools::getProductsOrder('by'), Tools::getProductsOrder('way')))

INSERT LINE
          ->setSortOrder(new SortOrder('product', Tools::getProductsOrder('by', 4), Tools::getProductsOrder('way', 1)))

................................

Regards,
Sergiy

**Comment by Kershoc** [comment]
I only concentrated on fixing the wrong value for Relevance since the defaults for the entire site can be set through the admin (Shop Parameters > Product Settings, Pagination Section).  

I'll leave whether product pages and search pages should share the same defaults to those with greater knowledge then I. :)  I'm sure it will be a spirited debate.

**Comment by sbordun** [comment]
Hi Kershoc,
My understanding that Search results should be sorted according to biggest "weight". It is different from Default Page sort on category page (you mentioned).
The weight could be set up through the admin : Shop Parameters -> Search, WEIGHT section.
As I understand - products with the biggest weight have to be on the top of the search (Order By Relevance DESC). 
But the source code has the bug and : products with the biggest weight are shown on the bottom of the search (Order By Relevance ASC) because the "relevance sorting for the search" was not part of your fix.  My fix fixed the Relevance sorting issue for the Product Search .

I proposed my changes as a separate pull request: "default AJAX Relevance search sort order is wrong #8458", but nobody reviewed them.

Do you know how to ask for a review ?

Regards ,
Sergiy

**Comment by Quetzacoalt91** [comment]
Can a product owner review that change? @vincentbz?
I wonder if there is a misunderstanding all around these "position" and "relevance" terms. From my understanding, lower the position, higher is the relevance.

I validated from a technical point of view, but I'm not sure there was an actual issue.

**Comment by sbordun** [comment]
Prestashop should sort products in the search results according to WEIGHT. You could find the WEIGHT explanation on Shop Parameters -> Search, WEIGHT section:  "We advise you to set a greater weight for words which appear in the name or reference of a product. This will allow the search results to be as precise and relevant as possible."

My understanding : products with the higher WEIGHT should be on the top. For a test I put 100 weight for Product Name and 1 for the rest.  But products with the search words in Products Name always appears at the bottom of the search (BUG!!!).  After looking into the source code I found that the search results are searched according to RELEVANCE ASC.  Changing to RELEVANCE DESC fixed the search WEIGHT issue.  #8458

@vincentbz , could you review the search sorting algorithm?

**Comment by vincentbz** [comment]
Hi,

@Quetzacoalt91 I'm checking it :)

**Comment by Quetzacoalt91** [comment]
I agree with your explanation @sbordun, however I'm not sure the "relevance" in the list is really related to the "weight". This could be a side effect which, in that case, solves your issue.

I may be wrong, but looking at the DOM when we open the list:
capture du 2017-11-06 14-55-31

Shows that we do not use the weight to order the products, but the position information:
capture du 2017-11-06 14-55-09

**Comment by sbordun** [comment]
hi @Quetzacoalt91,

I can't check the source right now, but when I looked into the source last time I found (i could be wrong) that weight is used to calculate dynamic position inside of the search results. And then the dynamic position was used in the relevance sort order.

changing to "relevance desc" for product search helped me to show products with the higher WEIGHT in front of the search results.

**Comment by Kershoc** [comment]
/classes/Search.php  in the Search::find() method.  There is a [subquery](https://github.com/PrestaShop/PrestaShop/blob/46479d24bff914850a1a0ef3b01e56bad5222e90/classes/Search.php#L230-L241) to select the weight as the field position.

Tools::getProductsOrder() will get whatever order/way is set in the admin (Shop Parameters > Product Settings, Pagination Section) unless you pass in a value to override the defaults.

This raises the question should the SearchController specify a hard coded default or should it just honor whats in the admin?  Honoring the admin settings are great as the shop owner can change the default to whatever they want and the search follows suite.  However it also means that if you want relevance to be the default that the admin settings will need to be changed to Default Order By 'Position in category' Default order method 'Descending' which may not be what you want if you have set category positions on your products.  Hard coding the default of course removes the ability to dynamically change the default.

Personally, I like that the defaults are configurable.  It's flexible and allows choice by the shop owner, that can be overridden by the shopper.  It's why I did not change the search defaults, and only fixed the order for Relevance.

Seems like there are a few options here.  
- Honor the admin default
- Hard code the default
- Independent order by settings for search pages and product pages
- Use a different field name for search sorted by relevance

**Comment by sbordun** [comment]
I believe that the order on Category pages should be different vs search order.
Logically: when user search for products - products with the higher WEIGHT has to be on the top of the search and it is different to Category page sort order (default: position inside of the category).

@Kershoc, we do not need separate settings for products pages sort order vs search sort order.
The WEIGHT settings should work like search order settings (products that collects the biggest SUM of word WEIGHTS should appear at the top of the search). Otherwise what is the sens of word WEIGHTS settings if it could be overridden by product page settings?

meanwhile: Prestashop 1.6 and 1.5 do not have this issue with the search order.  I tested the search and see that products with the higher Weight are shown on the top of the search results.  AT the same time products on the category page are shown according to position inside of the category.

I believe that programmers made a bug when rewriting the code for Prestashop 1.7.

I hope @vincentbz will review the search order code

**Comment by Quetzacoalt91** [comment]
Thank you everybody!

**Pull request merged**